diagnostic_tool/kinematics_cpp/kinematics.cpp

```diff
@@ -0,0 +1,443 @@
+/*
+ * Adapted from https://github.com/Le0nX/ModernRoboticsCpp
+ */
+#include <cmath>
+#include <vector>
+#include <Eigen/Dense>
+#include <pybind11/pybind11.h>
+#include <pybind11/eigen.h>
+
+namespace py = pybind11;
+
+#define M_PI           3.14159265358979323846  /* pi */
+
+namespace mr {
+
+/* Function: Find if the value is negligible enough to consider 0
+ * Inputs: value to be checked as a double
+ * Returns: Boolean of true-ignore or false-can't ignore
+ */
+    bool NearZero(const double val) {
+        return (std::abs(val) < .000001);
+    }
+
+/* Function: Returns the skew symmetric matrix representation of an angular velocity vector
+ * Input: Eigen::Vector3d 3x1 angular velocity vector
+ * Returns: Eigen::MatrixXd 3x3 skew symmetric matrix
+ */
+    Eigen::Matrix3d VecToso3(const Eigen::Vector3d &omg) {
+        Eigen::Matrix3d m_ret;
+        m_ret << 0, -omg(2), omg(1),
+                omg(2), 0, -omg(0),
+                -omg(1), omg(0), 0;
+        return m_ret;
+    }
+
+/*
+ * Function: Calculate the 6x6 matrix [adV] of the given 6-vector
+ * Input: Eigen::VectorXd (6x1)
+ * Output: Eigen::MatrixXd (6x6)
+ * Note: Can be used to calculate the Lie bracket [V1, V2] = [adV1]V2
+ */
+    Eigen::MatrixXd ad(Eigen::VectorXd V) {
+        Eigen::Matrix3d omgmat = VecToso3(Eigen::Vector3d(V(0), V(1), V(2)));
+
+        Eigen::MatrixXd result(6, 6);
+        result.topLeftCorner<3, 3>() = omgmat;
+        result.topRightCorner<3, 3>() = Eigen::Matrix3d::Zero(3, 3);
+        result.bottomLeftCorner<3, 3>() = VecToso3(Eigen::Vector3d(V(3), V(4), V(5)));
+        result.bottomRightCorner<3, 3>() = omgmat;
+        return result;
+    }
+
+/* Function: Returns a normalized version of the input vector
+ * Input: Eigen::MatrixXd
+ * Output: Eigen::MatrixXd
+ * Note: MatrixXd is used instead of VectorXd for the case of row vectors
+ * 		Requires a copy
+ *		Useful because of the MatrixXd casting
+ */
+    Eigen::MatrixXd Normalize(Eigen::MatrixXd V) {
+        V.normalize();
+        return V;
+    }

```

Comment:
In what situation would the user choose to call this function instead of just `v.normalize()`?

Reply:
Kept as is as much of the code is 3rd party source code, buuuut, this one is actually a bit silly.